Add pref64 support (RFC8781)

zajdee opened the pull request:
This pull requests adds support for the `pref64` option, as defined in RFC8781. Its configuration is part of the interface config section and has only two parameters: the `nat64prefix` and the optional `AdvValidLifetime`, the prefix valid lifetime (which should not be smaller than the `MaxRtrAdvInterval`).

```
interface eth0
{
        nat64prefix 64:ff9b::/96
        {
                AdvValidLifetime 1800;
        };
};
```

stappersg commented:
After seening additional commits, I marked this _merge request_ as _Work In Progress_.

robbat2 commented on gram.y:
```diff
@@ -478,6 +483,96 @@ v6addrlist_rasrcaddress	: IPV6ADDR ';'
+		{
+			if (nat64prefix) {
+
+				if (nat64prefix->AdvValidLifetime > 65528)
```
Please move this constant elsewhere, and document that it comes from the RFC

zajdee replied:
Done

robbat2 commented on gram.y:
```diff
@@ -478,6 +483,96 @@ v6addrlist_rasrcaddress	: IPV6ADDR ';'
+				if (nat64prefix->AdvValidLifetime > 65528)
+				{
+					flog(LOG_ERR, "AdvValidLifeTime must be "
+						"smaller or equal to 65528 in %s, line %d",
```
consume the value from the constant definition

zajdee replied:
Done

robbat2 commented on gram.y:
```diff
@@ -478,6 +483,96 @@ v6addrlist_rasrcaddress	: IPV6ADDR ';'
+			memset(&zeroaddr, 0, sizeof(zeroaddr));
+
+#ifndef HAVE_IFADDRS_H	// all-zeros is not a valid for NAT64 prefix
+			if (!memcmp($2, &zeroaddr, sizeof(struct in6_addr))) {
```
maybe hoist this to a function, because it's duplicated now?

robbat2 commented on gram.y:
```diff
@@ -478,6 +483,96 @@ v6addrlist_rasrcaddress	: IPV6ADDR ';'
+				flog(LOG_ERR, "invalid prefix length in %s, line %d", filename, num_lines);
+				ABORT;
+			}
+			if (
```
how about a switch statement instead?

zajdee replied:
Updated, thanks

robbat2 commented on gram.y:
```diff
@@ -478,6 +483,96 @@ v6addrlist_rasrcaddress	: IPV6ADDR ';'
+			}
+			nat64prefix->PrefixLen = $4;
+
+			if (3*(iface->MaxRtrAdvInterval) > DFLT_NAT64AdvValidLifetime) {
```
replace with:
`nat64prefix->AdvValidLifetime = min( DFLT_NAT64AdvValidLifetime, 3*(iface->MaxRtrAdvInterval));`

robbat2 commented on defaults.h:
```diff
@@ -64,6 +64,9 @@
 #define DFLT_DeprecatePrefixFlag 0
 #define DFLT_DecrementLifetimesFlag 0
 
+/* NAT64 prefix has an associated: */
```
reference RFC here?

zajdee replied:
RFC referenced

robbat2 commented:
- reviewed code and enabled the workflow after that review (sorry, have to check because of other projects where PRs contained malicious payloads that used CPU in workflow).
- Left comments on improvements, mostly style; functionality seems fine.

zajdee commented:
Hi @robbat2,
thank you for all the valuable inputs.

- I have updated the code according to your code review (also had to amend the original commits as I have used an incorrect e-mail address there) and also simplified it (`nat64prefix_init_defaults` now sets the default value of `AdvValidLifetime`).
- There also was an `unknown option 38` reported when `radvd` was running in debug mode, therefore I have skipped option 38 checking in `process.c`.
- Finally I have added option 38 parsing to `radvdump`.

Could you please re-run the workflows to ensure I didn't break things? Thanks.

stappersg commented:
> Could you please re-run the workflows to ensure I didn't break things?

Re-run triggered.

zajdee commented:
Hi gents,
I have further cleaned up the code and have nothing more to add/remove right now.
If there will be further changes needed based on your code review or as a result of the workflow checks, I will incorporate those.
Otherwise it's ready to be merged.

Thank you for your time reviewing this pull request.

stappersg commented:
Re-run triggered

stappersg commented:
> I have further cleaned up the code and have nothing more to add/remove right now.

I have removed the _Work In Progress marker_ from the title of this merge request.

robbat2 reviewed (approved):
LGTM. Thanks for your contribution.

Neustradamus commented:
2024-12-31: radvd 2.20 has been released with this nice PR, thanks to @zajdee :)
Never too late!
